### PR TITLE
MAINT: Upgrade to spin 0.13

### LINKED ISSUE #27432: `spin ipython` failing
Trying `spin ipython` fails with the following error:

```sh
$ spin ipython       
$ /home/adrin/micromamba/envs/numpy/bin/python3.12 vendored-meson/meson/meson.py compile -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/adrin/micromamba/envs/numpy/bin/ninja -C /home/adrin/Projects/gh/me/numpy/build
ninja: Entering directory `/home/adrin/Projects/gh/me/numpy/build'
[1/1] Generating numpy/generate-version with a custom command
Saving version to numpy/version.py
$ /home/adrin/micromamba/envs/numpy/bin/python3.12 vendored-meson/meson/meson.py install --only-changed -C build --destdir ../build-install

Traceback (most recent call last):
  File "/home/adrin/micromamba/envs/numpy/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adrin/micromamba/envs/numpy/lib/python3.12/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adrin/Projects/gh/me/numpy/.spin/cmds.py", line 682, in ipython
    ppath = meson._set_pythonpath()
            ^^^^^^^^^^^^^^^^^^^^^^^
TypeError: _set_pythonpath() missing 1 required positional argument: 'build_dir'

An internal error has occurred. Please file a bug report at

  https://github.com/scientific-python/spin

including the above traceback and the following information:

  spin: 0.12, package: numpy

Aborting.
```

`spin python` and `spin build` and `spin test` work as expected.

Installed package info:

<details>

```sh
$ mmamba list
List of packages in environment: "/home/adrin/micromamba/envs/numpy"

  Name                           Version       Build                Channel    
─────────────────────────────────────────────────────────────────────────────────
  _libgcc_mutex                  0.1           conda_forge          conda-forge
  _openmp_mutex                  4.5           2_gnu                conda-forge
  accessible-pygments            0.0.4         pyhd8ed1ab_0         conda-forge
  alabaster                      0.7.16        pyhd8ed1ab_0         conda-forge
  alsa-lib                       1.2.11        hd590300_1           conda-forge
  asttokens                      2.4.1         pyhd8ed1ab_0         conda-forge
  attr                           2.5.1         h166bdaf_1           conda-forge
  attrs                          23.2.0        pyh71513ae_0         conda-forge
  babel                          2.14.0        pyhd8ed1ab_0         conda-forge
  backports.zoneinfo             0.2.1         py312h7900ff3_8      conda-forge
  beautifulsoup4                 4.12.3        pyha770c72_0         conda-forge
  binutils                       2.40          h4852527_0           conda-forge
  binutils_impl_linux-64         2.40          ha885e6a_0           conda-forge
  binutils_linux-64              2.40          hdade7a5_3           conda-forge
  breathe                        4.35.0        pyhd8ed1ab_1         conda-forge
  brotli                         1.1.0         hd590300_1           conda-forge
  brotli-bin                     1.1.0         hd590300_1           conda-forge
  brotli-python                  1.1.0         py312h30efb56_1      conda-forge
  bzip2                          1.0.8         hd590300_5           conda-forge
  c-compiler                     1.7.0         hd590300_1           conda-forge
  ca-certificates                2024.2.2      hbcca054_0           conda-forge
  cairo                          1.18.0        h3faef2a_0           conda-forge
  ccache                         4.9.1         h1fcd64f_0           conda-forge
  certifi                        2024.2.2      pyhd8ed1ab_0         conda-forge
  cffi                           1.16.0        py312hf06ca03_0      conda-forge
  charset-normalizer             3.3.2         pyhd8ed1ab_0         conda-forge
  click                          8.1.7         unix_pyh707e725_0    conda-forge
  colorama                       0.4.6         pyhd8ed1ab_0         conda-forge
  compilers                      1.7.0         ha770c72_1           conda-forge
  contourpy                      1.2.1         py312h8572e83_0      conda-forge
  coverage                       7.5.1         py312h9a8786e_0      conda-forge
  cxx-compiler                   1.7.0         h00ab1b0_1           conda-forge
  cycler                         0.12.1        pyhd8ed1ab_0         conda-forge
  cython                         3.0.10        py312h30efb56_0      conda-forge
  dbus                           1.13.6        h5008d03_3           conda-forge
  decorator                      5.1.1         pyhd8ed1ab_0         conda-forge
  docutils                       0.17.1        py312h7900ff3_5      conda-forge
  doxygen                        1.10.0        h661eb56_0           conda-forge
  exceptiongroup                 1.2.0         pyhd8ed1ab_2         conda-forge
  execnet                        2.1.1         pyhd8ed1ab_0         conda-forge
  executing                      2.0.1         pyhd8ed1ab_0         conda-forge
  expat                          2.6.2         h59595ed_0           conda-forge
  font-ttf-dejavu-sans-mono      2.37          hab24e00_0           conda-forge
  font-ttf-inconsolata           3.000         h77eed37_0           conda-forge
  font-ttf-source-code-pro       2.038         h77eed37_0           conda-forge
  font-ttf-ubuntu                0.83          h77eed37_2           conda-forge
  fontconfig                     2.14.2        h14ed4e7_0           conda-forge
  fonts-conda-ecosystem          1             0                    conda-forge
  fonts-conda-forge              1             0                    conda-forge
  fonttools                      4.51.0        py312h98912ed_0      conda-forge
  fortran-compiler               1.7.0         heb67821_1           conda-forge
  freetype                       2.12.1        h267a509_2           conda-forge
  gcc                            12.3.0        h915e2ae_7           conda-forge
  gcc_impl_linux-64              12.3.0        h58ffeeb_7           conda-forge
  gcc_linux-64                   12.3.0        h6477408_3           conda-forge
  gettext                        0.22.5        h59595ed_2           conda-forge
  gettext-tools                  0.22.5        h59595ed_2           conda-forge
  gfortran                       12.3.0        h915e2ae_7           conda-forge
  gfortran_impl_linux-64         12.3.0        h1645026_7           conda-forge
  gfortran_linux-64              12.3.0        h617cb40_3           conda-forge
  gitdb                          4.0.11        pyhd8ed1ab_0         conda-forge
  gitpython                      3.1.43        pyhd8ed1ab_0         conda-forge
  glib                           2.80.2        hf974151_0           conda-forge
  glib-tools                     2.80.2        hb6ce0ca_0           conda-forge
  graphite2                      1.3.13        h59595ed_1003        conda-forge
  gst-plugins-base               1.24.3        h9ad1361_0           conda-forge
  gstreamer                      1.24.3        haf2f30d_0           conda-forge
  gxx                            12.3.0        h915e2ae_7           conda-forge
  gxx_impl_linux-64              12.3.0        h2a574ab_7           conda-forge
  gxx_linux-64                   12.3.0        h4a1b8e8_3           conda-forge
  harfbuzz                       8.4.0         h3d44ed6_0           conda-forge
  hypothesis                     6.101.0       pyha770c72_0         conda-forge
  icu                            73.2          h59595ed_0           conda-forge
  idna                           3.7           pyhd8ed1ab_0         conda-forge
  imagesize                      1.4.1         pyhd8ed1ab_0         conda-forge
  importlib-metadata             7.1.0         pyha770c72_0         conda-forge
  iniconfig                      2.0.0         pyhd8ed1ab_0         conda-forge
  ipython                        8.24.0        pyh707e725_0         conda-forge
  jedi                           0.19.1        pyhd8ed1ab_0         conda-forge
  jinja2                         3.1.4         pyhd8ed1ab_0         conda-forge
  kernel-headers_linux-64        2.6.32        he073ed8_17          conda-forge
  keyutils                       1.6.1         h166bdaf_0           conda-forge
  kiwisolver                     1.4.5         py312h8572e83_1      conda-forge
  krb5                           1.21.2        h659d440_0           conda-forge
  lame                           3.100         h166bdaf_1003        conda-forge
  lcms2                          2.16          hb7c19ff_0           conda-forge
  ld_impl_linux-64               2.40          h55db66e_0           conda-forge
  lerc                           4.0.0         h27087fc_0           conda-forge
  libasprintf                    0.22.5        h661eb56_2           conda-forge
  libasprintf-devel              0.22.5        h661eb56_2           conda-forge
  libblas                        3.9.0         22_linux64_openblas  conda-forge
  libbrotlicommon                1.1.0         hd590300_1           conda-forge
  libbrotlidec                   1.1.0         hd590300_1           conda-forge
  libbrotlienc                   1.1.0         hd590300_1           conda-forge
  libcap                         2.69          h0f662aa_0           conda-forge
  libcblas                       3.9.0         22_linux64_openblas  conda-forge
  libclang-cpp15                 15.0.7        default_h127d8a8_5   conda-forge
  libclang13                     18.1.5        default_h5d6823c_0   conda-forge
  libcups                        2.3.3         h4637d8d_4           conda-forge
  libdeflate                     1.20          hd590300_0           conda-forge
  libedit                        3.1.20191231  he28a2e2_2           conda-forge
  libevent                       2.1.12        hf998b51_1           conda-forge
  libexpat                       2.6.2         h59595ed_0           conda-forge
  libffi                         3.4.2         h7f98852_5           conda-forge
  libflac                        1.4.3         h59595ed_0           conda-forge
  libgcc-devel_linux-64          12.3.0        h0223996_107         conda-forge
  libgcc-ng                      13.2.0        h77fa898_7           conda-forge
  libgcrypt                      1.10.3        hd590300_0           conda-forge
  libgettextpo                   0.22.5        h59595ed_2           conda-forge
  libgettextpo-devel             0.22.5        h59595ed_2           conda-forge
  libgfortran-ng                 13.2.0        h69a702a_7           conda-forge
  libgfortran5                   13.2.0        hca663fb_7           conda-forge
  libglib                        2.80.2        hf974151_0           conda-forge
  libgomp                        13.2.0        h77fa898_7           conda-forge
  libgpg-error                   1.49          h4f305b6_0           conda-forge
  libhiredis                     1.0.2         h2cc385e_0           conda-forge
  libiconv                       1.17          hd590300_2           conda-forge
  libjpeg-turbo                  3.0.0         hd590300_1           conda-forge
  liblapack                      3.9.0         22_linux64_openblas  conda-forge
  libllvm15                      15.0.7        hb3ce162_4           conda-forge
  libllvm18                      18.1.5        hb77312f_0           conda-forge
  libnsl                         2.0.1         hd590300_0           conda-forge
  libogg                         1.3.4         h7f98852_1           conda-forge
  libopenblas                    0.3.27        pthreads_h413a1c8_0  conda-forge
  libopus                        1.3.1         h7f98852_1           conda-forge
  libpng                         1.6.43        h2797004_0           conda-forge
  libpq                          16.3          ha72fbe1_0           conda-forge
  libsanitizer                   12.3.0        hb8811af_7           conda-forge
  libsndfile                     1.2.2         hc60ed4a_1           conda-forge
  libsqlite                      3.45.3        h2797004_0           conda-forge
  libstdcxx-devel_linux-64       12.3.0        h0223996_107         conda-forge
  libstdcxx-ng                   13.2.0        hc0a3c3a_7           conda-forge
  libsystemd0                    255           h3516f8a_1           conda-forge
  libtiff                        4.6.0         h1dd3fc0_3           conda-forge
  libuuid                        2.38.1        h0b41bf4_0           conda-forge
  libvorbis                      1.3.7         h9c3ff4c_0           conda-forge
  libwebp-base                   1.4.0         hd590300_0           conda-forge
  libxcb                         1.15          h0b41bf4_0           conda-forge
  libxcrypt                      4.4.36        hd590300_1           conda-forge
  libxkbcommon                   1.7.0         h662e7e4_0           conda-forge
  libxml2                        2.12.6        h232c23b_2           conda-forge
  libzlib                        1.2.13        hd590300_5           conda-forge
  lz4-c                          1.9.4         hcb278e6_0           conda-forge
  markupsafe                     2.1.5         py312h98912ed_0      conda-forge
  matplotlib                     3.8.4         py312h7900ff3_0      conda-forge
  matplotlib-base                3.8.4         py312he5832f3_0      conda-forge
  matplotlib-inline              0.1.7         pyhd8ed1ab_0         conda-forge
  meson                          1.5.1         pyhd8ed1ab_1         conda-forge
  meson-python                   0.16.0        pyh0c530f3_0         conda-forge
  mpg123                         1.32.6        h59595ed_0           conda-forge
  munkres                        1.1.4         pyh9f0ad1d_0         conda-forge
  mypy                           1.7.1         py312h98912ed_1      conda-forge
  mypy_extensions                1.0.0         pyha770c72_0         conda-forge
  mysql-common                   8.3.0         hf1915f5_4           conda-forge
  mysql-libs                     8.3.0         hca2cd23_4           conda-forge
  ncurses                        6.5           h59595ed_0           conda-forge
  ninja                          1.12.1        h297d8ca_0           conda-forge
  nomkl                          1.0           h5ca1d4c_0           conda-forge
  nspr                           4.35          h27087fc_0           conda-forge
  nss                            3.100         hca3bf56_0           conda-forge
  numpy                          1.26.4        py312heda63a1_0      conda-forge
  numpydoc                       1.4.0         pyhd8ed1ab_1         conda-forge
  openblas                       0.3.27        pthreads_h7a3da1a_0  conda-forge
  openjpeg                       2.5.2         h488ebb8_0           conda-forge
  openssl                        3.3.0         hd590300_0           conda-forge
  packaging                      24.0          pyhd8ed1ab_0         conda-forge
  parso                          0.8.4         pyhd8ed1ab_0         conda-forge
  pcre2                          10.43         hcad00b1_0           conda-forge
  pexpect                        4.9.0         pyhd8ed1ab_0         conda-forge
  pickleshare                    0.7.5         py_1003              conda-forge
  pillow                         10.3.0        py312hdcec9eb_0      conda-forge
  pip                            24.0          pyhd8ed1ab_0         conda-forge
  pixman                         0.43.2        h59595ed_0           conda-forge
  pkg-config                     0.29.2        h36c2ea0_1008        conda-forge
  pluggy                         1.5.0         pyhd8ed1ab_0         conda-forge
  ply                            3.11          pyhd8ed1ab_2         conda-forge
  prompt-toolkit                 3.0.42        pyha770c72_0         conda-forge
  psutil                         5.9.8         py312h98912ed_0      conda-forge
  pthread-stubs                  0.4           h36c2ea0_1001        conda-forge
  ptyprocess                     0.7.0         pyhd3deb0d_0         conda-forge
  pulseaudio-client              17.0          hb77b528_0           conda-forge
  pure_eval                      0.2.2         pyhd8ed1ab_0         conda-forge
  pycodestyle                    2.8.0         pyhd8ed1ab_0         conda-forge
  pycparser                      2.22          pyhd8ed1ab_0         conda-forge
  pydata-sphinx-theme            0.13.3        pyhd8ed1ab_0         conda-forge
  pygments                       2.18.0        pyhd8ed1ab_0         conda-forge
  pyparsing                      3.1.2         pyhd8ed1ab_0         conda-forge
  pyproject-metadata             0.8.0         pyhd8ed1ab_0         conda-forge
  pyqt                           5.15.9        py312h949fe66_5      conda-forge
  pyqt5-sip                      12.12.2       py312h30efb56_5      conda-forge
  pysocks                        1.7.1         pyha2e5f31_6         conda-forge
  pytest                         8.2.0         pyhd8ed1ab_0         conda-forge
  pytest-cov                     5.0.0         pyhd8ed1ab_0         conda-forge
  pytest-xdist                   3.5.0         pyhd8ed1ab_0         conda-forge
  python                         3.12.3        hab00c5b_0_cpython   conda-forge
  python-dateutil                2.9.0         pyhd8ed1ab_0         conda-forge
  python_abi                     3.12          4_cp312              conda-forge
  pytz                           2024.1        pyhd8ed1ab_0         conda-forge
  qt-main                        5.15.8        hc9dc06e_21          conda-forge
  readline                       8.2           h8228510_1           conda-forge
  requests                       2.31.0        pyhd8ed1ab_0         conda-forge
  scipy                          1.13.0        py312hc2bc53b_1      conda-forge
  setuptools                     69.5.1        pyhd8ed1ab_0         conda-forge
  sip                            6.7.12        py312h30efb56_0      conda-forge
  six                            1.16.0        pyh6c4a22f_0         conda-forge
  smmap                          5.0.0         pyhd8ed1ab_0         conda-forge
  snowballstemmer                2.2.0         pyhd8ed1ab_0         conda-forge
  sortedcontainers               2.4.0         pyhd8ed1ab_0         conda-forge
  soupsieve                      2.5           pyhd8ed1ab_1         conda-forge
  sphinx                         5.3.0         pyhd8ed1ab_0         conda-forge
  sphinx-design                  0.5.0         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-applehelp        1.0.8         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-devhelp          1.0.6         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-htmlhelp         2.0.5         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-jsmath           1.0.1         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-qthelp           1.0.7         pyhd8ed1ab_0         conda-forge
  sphinxcontrib-serializinghtml  1.1.10        pyhd8ed1ab_0         conda-forge
  stack_data                     0.6.2         pyhd8ed1ab_0         conda-forge
  sysroot_linux-64               2.12          he073ed8_17          conda-forge
  tk                             8.6.13        noxft_h4845f30_101   conda-forge
  toml                           0.10.2        pyhd8ed1ab_0         conda-forge
  tomli                          2.0.1         pyhd8ed1ab_0         conda-forge
  tornado                        6.4           py312h98912ed_0      conda-forge
  traitlets                      5.14.3        pyhd8ed1ab_0         conda-forge
  typing_extensions              4.11.0        pyha770c72_0         conda-forge
  tzdata                         2024a         h0c530f3_0           conda-forge
  urllib3                        2.2.1         pyhd8ed1ab_0         conda-forge
  wcwidth                        0.2.13        pyhd8ed1ab_0         conda-forge
  wheel                          0.43.0        pyhd8ed1ab_1         conda-forge
  xcb-util                       0.4.0         hd590300_1           conda-forge
  xcb-util-image                 0.4.0         h8ee46fc_1           conda-forge
  xcb-util-keysyms               0.4.0         h8ee46fc_1           conda-forge
  xcb-util-renderutil            0.3.9         hd590300_1           conda-forge
  xcb-util-wm                    0.4.1         h8ee46fc_1           conda-forge
  xkeyboard-config               2.41          hd590300_0           conda-forge
  xorg-kbproto                   1.0.7         h7f98852_1002        conda-forge
  xorg-libice                    1.1.1         hd590300_0           conda-forge
  xorg-libsm                     1.2.4         h7391055_0           conda-forge
  xorg-libx11                    1.8.9         h8ee46fc_0           conda-forge
  xorg-libxau                    1.0.11        hd590300_0           conda-forge
  xorg-libxdmcp                  1.1.3         h7f98852_0           conda-forge
  xorg-libxext                   1.3.4         h0b41bf4_2           conda-forge
  xorg-libxrender                0.9.11        hd590300_0           conda-forge
  xorg-renderproto               0.11.1        h7f98852_1002        conda-forge
  xorg-xextproto                 7.3.0         h0b41bf4_1003        conda-forge
  xorg-xf86vidmodeproto          2.3.1         h7f98852_1002        conda-forge
  xorg-xproto                    7.0.31        h7f98852_1007        conda-forge
  xz                             5.2.6         h166bdaf_0           conda-forge
  zipp                           3.17.0        pyhd8ed1ab_0         conda-forge
  zlib                           1.2.13        hd590300_5           conda-forge
  zstd                           1.5.6         ha6fb4c9_0           conda-forge


$ pip freeze
accessible-pygments @ file:///home/conda/feedstock_root/build_artifacts/accessible-pygments_1679583834850/work
alabaster @ file:///home/conda/feedstock_root/build_artifacts/alabaster_1704848697227/work
asttokens @ file:///home/conda/feedstock_root/build_artifacts/asttokens_1698341106958/work
attrs @ file:///home/conda/feedstock_root/build_artifacts/attrs_1704011227531/work
Babel @ file:///home/conda/feedstock_root/build_artifacts/babel_1702422572539/work
beautifulsoup4 @ file:///home/conda/feedstock_root/build_artifacts/beautifulsoup4_1705564648255/work
breathe @ file:///home/conda/feedstock_root/build_artifacts/breathe_1688446161072/work
Brotli @ file:///home/conda/feedstock_root/build_artifacts/brotli-split_1695989787169/work
certifi @ file:///home/conda/feedstock_root/build_artifacts/certifi_1707022139797/work/certifi
cffi @ file:///home/conda/feedstock_root/build_artifacts/cffi_1696001721842/work
charset-normalizer @ file:///home/conda/feedstock_root/build_artifacts/charset-normalizer_1698833585322/work
click @ file:///home/conda/feedstock_root/build_artifacts/click_1692311806742/work
colorama @ file:///home/conda/feedstock_root/build_artifacts/colorama_1666700638685/work
contourpy @ file:///home/conda/feedstock_root/build_artifacts/contourpy_1712429918028/work
coverage @ file:///home/conda/feedstock_root/build_artifacts/coverage_1714846632857/work
cycler @ file:///home/conda/feedstock_root/build_artifacts/cycler_1696677705766/work
Cython @ file:///home/conda/feedstock_root/build_artifacts/cython_1711833953512/work
decorator @ file:///home/conda/feedstock_root/build_artifacts/decorator_1641555617451/work
docutils @ file:///home/conda/feedstock_root/build_artifacts/docutils_1713293345733/work
exceptiongroup @ file:///home/conda/feedstock_root/build_artifacts/exceptiongroup_1704921103267/work
execnet @ file:///home/conda/feedstock_root/build_artifacts/execnet_1712591832280/work
executing @ file:///home/conda/feedstock_root/build_artifacts/executing_1698579936712/work
fonttools @ file:///home/conda/feedstock_root/build_artifacts/fonttools_1712344557282/work
gitdb @ file:///home/conda/feedstock_root/build_artifacts/gitdb_1697791558612/work
GitPython @ file:///home/conda/feedstock_root/build_artifacts/gitpython_1711991025291/work
hypothesis @ file:///home/conda/feedstock_root/build_artifacts/hypothesis_1715587946681/work
idna @ file:///home/conda/feedstock_root/build_artifacts/idna_1713279365350/work
imagesize @ file:///home/conda/feedstock_root/build_artifacts/imagesize_1656939531508/work
importlib_metadata @ file:///home/conda/feedstock_root/build_artifacts/importlib-metadata_1710971335535/work
iniconfig @ file:///home/conda/feedstock_root/build_artifacts/iniconfig_1673103042956/work
ipython @ file:///home/conda/feedstock_root/build_artifacts/ipython_1715263367085/work
jedi @ file:///home/conda/feedstock_root/build_artifacts/jedi_1696326070614/work
Jinja2 @ file:///home/conda/feedstock_root/build_artifacts/jinja2_1715127149914/work
kiwisolver @ file:///home/conda/feedstock_root/build_artifacts/kiwisolver_1695379925569/work
MarkupSafe @ file:///home/conda/feedstock_root/build_artifacts/markupsafe_1706899920239/work
matplotlib @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-suite_1712605872637/work
matplotlib-inline @ file:///home/conda/feedstock_root/build_artifacts/matplotlib-inline_1713250518406/work
meson @ file:///home/conda/feedstock_root/build_artifacts/meson_1722578149419/work
meson-python @ file:///home/conda/feedstock_root/build_artifacts/meson-python_1713362482020/work
munkres==1.1.4
mypy @ file:///home/conda/feedstock_root/build_artifacts/mypy-split_1702295589952/work
mypy-extensions @ file:///home/conda/feedstock_root/build_artifacts/mypy_extensions_1675543315189/work
numpy==2.1.1
numpydoc @ file:///home/conda/feedstock_root/build_artifacts/numpydoc_1658207752219/work
packaging @ file:///home/conda/feedstock_root/build_artifacts/packaging_1710075952259/work
pandas==2.2.2
parso @ file:///home/conda/feedstock_root/build_artifacts/parso_1712320355065/work
pexpect @ file:///home/conda/feedstock_root/build_artifacts/pexpect_1706113125309/work
pickleshare @ file:///home/conda/feedstock_root/build_artifacts/pickleshare_1602536217715/work
pillow @ file:///home/conda/feedstock_root/build_artifacts/pillow_1712154464349/work
pluggy @ file:///home/conda/feedstock_root/build_artifacts/pluggy_1713667077545/work
ply @ file:///home/conda/feedstock_root/build_artifacts/ply_1712242996588/work
polars==1.7.1
prompt-toolkit @ file:///home/conda/feedstock_root/build_artifacts/prompt-toolkit_1702399386289/work
psutil @ file:///home/conda/feedstock_root/build_artifacts/psutil_1705722396628/work
ptyprocess @ file:///home/conda/feedstock_root/build_artifacts/ptyprocess_1609419310487/work/dist/ptyprocess-0.7.0-py2.py3-none-any.whl
pure-eval @ file:///home/conda/feedstock_root/build_artifacts/pure_eval_1642875951954/work
pycodestyle @ file:///home/conda/feedstock_root/build_artifacts/pycodestyle_1633982426610/work
pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1711811537435/work
pydata-sphinx-theme==0.13.3
Pygments @ file:///home/conda/feedstock_root/build_artifacts/pygments_1714846767233/work
pyparsing @ file:///home/conda/feedstock_root/build_artifacts/pyparsing_1709721012883/work
pyproject-metadata @ file:///home/conda/feedstock_root/build_artifacts/pyproject-metadata_1713446809116/work
PyQt5==5.15.9
PyQt5-sip==12.12.2
PySocks @ file:///home/conda/feedstock_root/build_artifacts/pysocks_1661604839144/work
pytest @ file:///home/conda/feedstock_root/build_artifacts/pytest_1714308352747/work
pytest-cov @ file:///home/conda/feedstock_root/build_artifacts/pytest-cov_1711411024363/work
pytest-xdist @ file:///home/conda/feedstock_root/build_artifacts/pytest-xdist_1700592942746/work
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1709299778482/work
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1706886791323/work
requests @ file:///home/conda/feedstock_root/build_artifacts/requests_1684774241324/work
scipy @ file:///home/conda/feedstock_root/build_artifacts/scipy-split_1714794086036/work/dist/scipy-1.13.0-cp312-cp312-linux_x86_64.whl#sha256=85fcf9f3c463027dde642b83ac2070caf3f26445076f537d47ed38c6d167ec75
setuptools==69.5.1
sip @ file:///home/conda/feedstock_root/build_artifacts/sip_1697300425834/work
six @ file:///home/conda/feedstock_root/build_artifacts/six_1620240208055/work
smmap @ file:///home/conda/feedstock_root/build_artifacts/smmap_1634310307496/work
snowballstemmer @ file:///home/conda/feedstock_root/build_artifacts/snowballstemmer_1637143057757/work
sortedcontainers @ file:///home/conda/feedstock_root/build_artifacts/sortedcontainers_1621217038088/work
soupsieve @ file:///home/conda/feedstock_root/build_artifacts/soupsieve_1693929250441/work
Sphinx @ file:///home/conda/feedstock_root/build_artifacts/sphinx_1665915552897/work
sphinx_design @ file:///home/conda/feedstock_root/build_artifacts/sphinx-design_1690472536755/work
sphinxcontrib-applehelp @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-applehelp_1705126298355/work
sphinxcontrib-devhelp @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-devhelp_1705126010477/work
sphinxcontrib-htmlhelp @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-htmlhelp_1705118152391/work
sphinxcontrib-jsmath @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-jsmath_1691604704163/work
sphinxcontrib-qthelp @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-qthelp_1705126152907/work
sphinxcontrib-serializinghtml @ file:///home/conda/feedstock_root/build_artifacts/sphinxcontrib-serializinghtml_1705118225549/work
spin==0.12
stack-data @ file:///home/conda/feedstock_root/build_artifacts/stack_data_1669632077133/work
toml @ file:///home/conda/feedstock_root/build_artifacts/toml_1604308577558/work
tomli @ file:///home/conda/feedstock_root/build_artifacts/tomli_1644342247877/work
tornado @ file:///home/conda/feedstock_root/build_artifacts/tornado_1708363096407/work
traitlets @ file:///home/conda/feedstock_root/build_artifacts/traitlets_1713535121073/work
typing_extensions @ file:///home/conda/feedstock_root/build_artifacts/typing_extensions_1712329955671/work
tzdata==2024.1
urllib3 @ file:///home/conda/feedstock_root/build_artifacts/urllib3_1708239446578/work
wcwidth @ file:///home/conda/feedstock_root/build_artifacts/wcwidth_1704731205417/work
wheel==0.43.0
zipp @ file:///home/conda/feedstock_root/build_artifacts/zipp_1695255097490/work
```

</details>

### PR DESCRIPTION
**Pull request by stefanv**
Closes gh-27432

### CODE REVIEW

**Comment by mattip** [comment]
The environment.yml needs updating https://github.com/numpy/numpy/blob/a17906094972cecfb832cc02141dfde71faa182a/environment.yml#L20

**Comment by mattip** [comment]
Why not wait for an official conda-forge release? Is there something in spin 0.13 that we require for NumPy?

**Comment by stefanv** [comment]
Sure, this was in response to a request by @ngoldbaum, but I'm happy to wait.

**Comment by ngoldbaum** [comment]
I didn’t look at the full diff, let’s also wait for a pypi release to merge this

**Comment by stefanv** [comment]
Just to be clear: this PR is to test `spin` with `numpy` in anticipation of the imminent 0.13 release. It also refactors `numpy`'s custom commands which, because of a new [extending mechanism](https://github.com/scientific-python/spin/pull/248), should now be much simpler.

It looks like CI is passing, but I want to make sure that we do not accidentally break any current developer patterns with the refactoring, so if y'all want to try out spin==0.13rc0 locally, that'd be great.

File: .spin/cmds.py [hunk 1]
```diff
@@ -611,7 +442,7 @@ def bench(ctx, tests, compare, verbose, quick, commits):
 
         meson._set_pythonpath()
```
Comment (ngoldbaum) [inline .spin/cmds.py:443]: `spin bench` is dying here on CI:

```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/hostedtoolcache/Python/3.10.15/x64/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/runner/work/numpy/numpy/.spin/cmds.py", line 443, in bench
    meson._set_pythonpath()
TypeError: _set_pythonpath() missing 1 required positional argument: 'build_dir'
```

**Comment by ngoldbaum** [comment]
I'm able to build numpy with free-threaded Python on my Windows developer machine with this, bypassing the issue I ran into related to how spin used to pass --prefix. None of my normal development workflows seem to be affected.

**Comment by ngoldbaum** [comment]
> If you need it to be, we could also expose it as a SPIN_BUILD_PREFIX or similar env var.

Not needed. The problem was that passing `--prefix=/usr` was hardcoded and for some reason only in some situations this caused meson to immediately exit with an error. Because it was hardcoded it couldn't be overrided without patching spin and numpy's spin config.

This PR finally closes that loop and patches numpy's spin config.

File: environment.yml [hunk 2]
```diff
@@ -16,8 +16,8 @@ dependencies:
   - ninja
   - pkg-config
   - meson-python
-  - pip
-  - spin=0.8 # Unpin when spin 0.9.1 is released
+  - pip:
+      - spin==0.13
```
Comment (seberg) [inline environment.yml:20]: Sorry, I forgot that this might have existed...  This is unnecessary of course.

**Pull request merged**